### ISSUE TITLE
simulating Atari 7800 startup process

### ISSUE THREAD

**Issue by SvOlli** [stella-emu/stella#318]
Last weekend, I stumbled across a very interesting problem: when using an F4 cartridge in a 7800, the bootrom triggered hotspots on the cartridge that left it in a different state than the 2600. It would be a nice feature to test this in Stella as well, since this can't be even tested with SD based cartridges.

**Comment by thrust26** [comment]
Sorry, I don't get you. Can you explain again?

**Comment by SvOlli** [comment]
Lets take it from the observer's view:
When an F4 cartridge starts in a 2600, the starting bank is always the first one.
When the same cartridge is run in a 7800, first the bootrom starts up and seems to access some hotspots on the cartridge. I assume it's done by the routine that tests for a 7800 cartridge. This leaves the cartridge with a starting bank as of "5 ($FF9)" before going to 2600 mode.

It would be nice to have an option that would simulate this behaviour, as that kind of problem only occurs on production hardware.

**Comment by thrust26** [comment]
Are you sure about the 2600 always starting at the first bank? From my observation the bank is random.

And this can be emulated using the developer options.

**Comment by SvOlli** [comment]
Random is true for (at least some implementations of) F6 and F8. F4 is always bank 0. [Scene Demos 6-in-1](http://xayax.net/nonxayax/scene_demos_6-in-1/) will show that: it runs one of five different demo, according to the starting bank. On "real" (CPLD) hardware it always starts at bank 0, as well as on a Melody board.

**Comment by thrust26** [comment]
OK, to emulate this, we need the 7800 boot ROM and then emulate it before starting the cartridge, correct?

**Comment by thrust26** [comment]
Maybe executing only this code will do the trick in your case
```
          LDA     $FFF8                  ;CHECK FOR REGION VERIFICATI...
          ORA     #REGION                ; = $FE
          CMP     #$FF
          BNE     BADCART
          LDA     $FFF9                  ;SEE IF MARIA SIGNATURE EXIS...
          AND     #$0B                   ;$07 OR $03 VALID
          CMP     #$03
          BNE     BADCART
          LDA     $FFF9                  ;GET BOTTOM OF CART ADDRESS
          AND     #$F0
          STA     CARTBOTM
          STA     CSCMOD0+2-CODEDIF      ;SET UP FOR START OF CHECKSU...
          CMP     #$40                   ;MAKE SURE IT IS NOT TOO LOW
          BCC     BADCART
;         SEC
          SBC     #$01                   ;MAKE SURE WE GET FENCEPOST ...
          CMP     $FFFD                  ;MAKE SURE START VECTOR WITH...
          BCS     BADCART
```
There are a lot previous hotspot accesses, but this is the only code which branches to BADCART (= 2600 mode).  And this would fail for other bankswitching types.

You told me in a private mail, that the bank is always 5, so I suppose the 2nd branch is taken here. Which means Stella gets bit 0 as 1 when reading from $FFF8. I don't have the time to check the results when reading a hotspot, someone should check the code.

**Comment by thrust26** [comment]
I know. But I am pretty sure the latter will not happen anytime soon.

**Comment by SvOlli** [comment]
I helped myself by using the manual bankswitching and the "reset" command from the debugger to test my workaround. It should be enough for my current problem. It's up to you to decide if you want to close the issue or keep it open for a later feature implementation.

**Comment by thrust26** [comment]
@sa666666 Shall we cancel it?

**Comment by sa666666** [comment]
Unless it's something that you think can be easily added, I say it's outside the scope of Stella and should be cancelled.

Basically, I don't personally have the time to implement it, so if you do, then fine, go ahead :smiling_imp: Otherwise we should kill it.

**Comment by thrust26** [comment]
Yes, doesn't sound feasible within our given resources.
